Multiple Inputs: update the Model module

amal-ghamdi opened the pull request:
closes #605 

Notes about reviewing: see issue description

nabriis reviewed (commented):
**From live review of code**
1) Printing looks a bit strange for Model. Maybe use specific case for ProductGeometry?
2) Can you check that if we pass in CUQIArray but say its funval/parameters what happens? Works as intended?
3) Related to stacking. Do we have good docstring to explain this option? Should we ask for user to provide e.g. "is_stacked".
4) What about Samples in the Dict? Why is the stacked samples not support?
5) Check that given RVs x, y, z, that F(x, y, z) works.
6) Look into gradient as None. Do we need  it now? Seems to change behavior. Maybe we return error?
7) Gradient as stacked. Look into "is_stacked" parameter.
8) Look into stacked output.
9) Look into if Model should be two classes or one?

amal-ghamdi commented:
@nabriis , if you got a chance, it will be great if you can provide your review on the PR. I am also happy to go over parts that are not clear in a meeting :). Many thanks!

chaozg commented on cuqi/model/_model.py:
```diff
@@ -87,64 +92,140 @@ def jacobian(x): # Can use "x" or "wrt" as the input argument name
+        def forward(x, y):
+            return 10 * y - 10 * x**3 + 5 * x**2 + 6 * x
+    
+        def jacobian(x, y):
```
What is the signature of the gradient function like?

chaozg commented on cuqi/model/_model.py:
```diff
@@ -160,334 +241,808 @@ def range_dim(self):
+                "The forward operator input is specified by more than one argument. This is only supported for domain geometry of type tuple of cuqi.geometry.Geometry objects."
+            )
+
+    def _check_correct_gradient_jacobian_form(self, func, func_type):
```
Beside signature, I wonder if it makes sense to also check the return type and possibly size.

chaozg commented on cuqi/model/_model.py:
```diff
@@ -160,334 +241,808 @@ def range_dim(self):
 
-    def __call__(self, *args, **kwargs):
-        return self.forward(*args, **kwargs)
+    def _handle_distributions(self, kwargs):
```
Could we have a more expressive name? I mean it's not clear what this function is doing from its name (even with together with the docstring)

amal-ghamdi replied:
I updated the name, let me know if you have other suggestions.

chaozg reviewed (commented):
Tiny comments on `init` and `forward`

nabriis commented on cuqi/model/_model.py:
```diff
@@ -87,64 +92,140 @@ def jacobian(x): # Can use "x" or "wrt" as the input argument name
+                *value)
+        # Remaining cases, forward model with single input
+        elif isinstance(value, tuple) and len(value) == 2:
+            self._domain_geometry = _DefaultGeometry2D(value)
```
This may need an overhaul depending if single or multiple inputs

nabriis commented on cuqi/model/_model.py:
```diff
@@ -160,334 +241,808 @@ def range_dim(self):
+        """Private function that checks if the domain geometry of the model is
+        consistent with the forward operator."""
+
+        non_default_args = cuqi.utilities.get_non_default_args(self._forward_func)
```
We have a property for this. Why repeat it here?

amal-ghamdi replied:
updated as suggested through a new property `number_of_inputs`

nabriis commented on cuqi/model/_model.py:
```diff
@@ -160,334 +241,808 @@ def range_dim(self):
+        consistent with the forward operator."""
+
+        non_default_args = cuqi.utilities.get_non_default_args(self._forward_func)
+        if (
+            not isinstance(
+                self.domain_geometry, cuqi.experimental.geometry._ProductGeometry
+            )
+            and len(non_default_args) > 1
+        ):
+            raise ValueError(
+                "The forward operator input is specified by more than one argument. This is only supported for domain geometry of type tuple of cuqi.geometry.Geometry objects."
+            )
```
This is rather besides the point of this PR, but the formatting here using (i think) black makes the review somewhat harder in my opinion.

amal-ghamdi replied:
I know 🙂, I left it as is for now which is auto-formatted.

nabriis commented on cuqi/model/_model.py:
```diff
@@ -160,334 +241,808 @@ def range_dim(self):
         # in case of operations such as "@" that can be interpreted as both __matmul__ and __rmatmul__
         # the operation may be delegated to the Node class.
-        if isinstance(x, cuqi.experimental.algebra.Node):
+        if len(args)== 1 and isinstance(args[0], cuqi.experimental.algebra.Node):
```
```suggestion
        if len(args) == 1 and isinstance(args[0], cuqi.experimental.algebra.Node):
```

nabriis commented on cuqi/model/_model.py:
```diff
@@ -160,334 +241,808 @@ def range_dim(self):
-        
+        # extract args from kwargs
+        args = list(kwargs.values())
+        if len(args)== 1 and isinstance(args[0], cuqi.experimental.algebra.RandomVariable):
```
```suggestion
        if len(args) == 1 and isinstance(args[0], cuqi.experimental.algebra.RandomVariable):
```

nabriis reviewed (commented):
Thanks @amal-ghamdi. I found some time to do a review over the holiday. I did not dive into the stacked options though.

Overall the feature seems very close to complete. The comprehensive testing framework is really nice! I had some small observations + the following comments:


1) I suggest updating the repr of model now that we can have geometries that can span multiple lines. To avoid this:
```bash
CUQI Model: _ProductGeometry(
	Discrete(1,)
	Discrete(1,)
) -> _DefaultGeometry1D(1,).
    Forward parameters: ['x', 'y'].
```
2) There seems to be an input discrepancy we need to look into. For models with multiple inputs one might expect to be able to do this:
```python
def forward(x, y):
    return 10 * y - 10 * x**3 + 5 * x**2 + 6 * x

def jacobian(x, y):
    return (np.array([[-30 * x**2 + 10 * x + 6]]), np.array([[10]]))

model = Model(
    forward,
    range_geometry=1,
    domain_geometry=(1, 1),
    jacobian=jacobian,
)
```
however this is not possible.
3) For this case, I notice the behavior is different than for 1 variable. Namely, we "lost" the model information because the parameter names should only have been changed.
```python
import numpy as np
import cuqi
from cuqi.model import Model
from cuqi.geometry import Discrete

def forward(x, y):
    return 10 * y - 10 * x**3 + 5 * x**2 + 6 * x

def jacobian(x, y):
    return (np.array([[-30 * x**2 + 10 * x + 6]]), np.array([[10]]))

model = Model(
    forward,
    range_geometry=1,
    domain_geometry=(Discrete(1), Discrete(1)),
    jacobian=jacobian,
)

z = cuqi.experimental.algebra.RandomVariable(cuqi.distribution.Normal(0,1))
v = cuqi.experimental.algebra.RandomVariable(cuqi.distribution.Normal(0,1))

model(z,v)
```

chaozg commented on cuqi/model/_model.py:
```diff
@@ -160,334 +241,808 @@ def range_dim(self):
+
+        # Convert to function values
+        for i, (k, v) in enumerate(kwargs.items()):
+            # Use CUQIarray funvals if geometry is consistent
+            if isinstance(v, CUQIarray) and v.geometry == geometries[i]:
+                kwargs[k] = v.funvals
+            # Otherwise we use the geometry par2fun method
+            elif is_par[i] and v is not None:
+                kwargs[k] = geometries[i].par2fun(v)
```
The comments seem not sync with the code. I mean 'Otherwise' with `elif`.

amal-ghamdi replied:
I made few updates to clarify this

amal-ghamdi commented on cuqi/model/_model.py:
```diff
@@ -160,334 +241,808 @@ def range_dim(self):
+            geometry=func_range_geometry, to_CUQIarray=is_CUQIarray, **{"out": out}
+        )["out"]
+
+    def _handle_samples(self, func=None, fwd=True, **kwargs):
```
I also updated the name here

amal-ghamdi commented:
> Thanks @amal-ghamdi. I found some time to do a review over the holiday. I did not dive into the stacked options though.
> 
> Overall the feature seems very close to complete. The comprehensive testing framework is really nice! I had some small observations + the following comments:
> 
> 1. I suggest updating the repr of model now that we can have geometries that can span multiple lines. To avoid this:
> 
> ```shell
> CUQI Model: _ProductGeometry(
> 	Discrete(1,)
> 	Discrete(1,)
> ) -> _DefaultGeometry1D(1,).
>     Forward parameters: ['x', 'y'].
> ```
> 
> 2. There seems to be an input discrepancy we need to look into. For models with multiple inputs one might expect to be able to do this:
> 
> ```python
> def forward(x, y):
>     return 10 * y - 10 * x**3 + 5 * x**2 + 6 * x
> 
> def jacobian(x, y):
>     return (np.array([[-30 * x**2 + 10 * x + 6]]), np.array([[10]]))
> 
> model = Model(
>     forward,
>     range_geometry=1,
>     domain_geometry=(1, 1),
>     jacobian=jacobian,
> )
> ```
> 
> however this is not possible. 3) For this case, I notice the behavior is different than for 1 variable. Namely, we "lost" the model information because the parameter names should only have been changed.
> 
> ```python
> import numpy as np
> import cuqi
> from cuqi.model import Model
> from cuqi.geometry import Discrete
> 
> def forward(x, y):
>     return 10 * y - 10 * x**3 + 5 * x**2 + 6 * x
> 
> def jacobian(x, y):
>     return (np.array([[-30 * x**2 + 10 * x + 6]]), np.array([[10]]))
> 
> model = Model(
>     forward,
>     range_geometry=1,
>     domain_geometry=(Discrete(1), Discrete(1)),
>     jacobian=jacobian,
> )
> 
> z = cuqi.experimental.algebra.RandomVariable(cuqi.distribution.Normal(0,1))
> v = cuqi.experimental.algebra.RandomVariable(cuqi.distribution.Normal(0,1))
> 
> model(z,v)
> ```


Really good points, 

**regarding the random variables**, when I finished this PR, I meant to add this issue as a separate issue but did not document that in here (only documented in personal notes). So now I just fixed it here with this PR instead, which makes sense. 

**I also updated the accepted domain geometry type** to support the example you provided along with updating the documentation of domain and range geometry. You can see the documentation and the added test `test_setting_domain_and_range_geometry` in `test_model.py` for the accepted format

**I updated the formatting**, single input:
```
CUQI Model: _DefaultGeometry1D(1,) -> _DefaultGeometry1D(1,).
    Forward parameters: ['x'].
```

Multiple inputs:
```
CUQI Model: _ProductGeometry(
                _DefaultGeometry1D(1,)
                _DefaultGeometry1D(1,)
            ) -> _DefaultGeometry1D(1,).
    Forward parameters: ['x', 'y'].
```

amal-ghamdi commented:
Many thanks @chaozg and @nabriis for your review! I know this is unusually big PR, so many thanks for taking the time to go through it. I addressed your points and I am requesting your review again.

@nabriis 
I added a generalization for the case of random variable internal nodes (from one input to multiple input). Let me know if this makes sence (line 771 in model):
```
        # the operation may be delegated to the Node class.
        elif any(isinstance(args_i, cuqi.experimental.algebra.Node) for args_i in args):
            return NotImplemented
```

nabriis commented:
> Many thanks @chaozg and @nabriis for your review! I know this is unusually big PR, so many thanks for taking the time to go through it. I addressed your points and I am requesting your review again.
> 
> @nabriis I added a generalization for the case of random variable internal nodes (from one input to multiple input). Let me know if this makes sence (line 771 in model):
> 
> ```
>         # the operation may be delegated to the Node class.
>         elif any(isinstance(args_i, cuqi.experimental.algebra.Node) for args_i in args):
>             return NotImplemented
> ```

Thanks @amal-ghamdi. Yes I think this is the right update.

nabriis reviewed (approved):
Thanks @amal-ghamdi  for reacting to my changes. The updates LGTM! Let us get this one merged ASAP!

chaozg commented on cuqi/model/_model.py:
```diff
@@ -160,341 +264,841 @@ def range_dim(self):
+                                                                  grad,
+                                                                  grad_is_par,
+                                                                  to_CUQIarray):
+        """ Private function that applies the chain rule to account for the
```
It will be good to state/hint in the docstring or somewhere else, like next to corresponding lines of codes, that in the end we are computing the gradient with respect to the "parameter" values, not the "function" values.

amal-ghamdi replied:
done

chaozg commented on cuqi/model/_model.py:
```diff
@@ -599,19 +1249,35 @@ def shift(self, value):
+        args = list(kwargs.values())
+        # if model has _original_non_default_args, we use it to replace the
+        # kwargs keys so that it matches self._linear_operator signature
+        if hasattr(self, '_original_non_default_args'):
```
Just a reminder from our discussion that in case this check is also needed in other functions

amal-ghamdi replied:
Thank you @chaozg, I doubled checked this case and realize we do not need to do something about it. Because changing the default argument names should only affect the parameter names for forward and gradient but not adjoint.

chaozg reviewed (commented):
@amal-ghamdi, I just left two comments from our review discussion

chaozg commented on tests/test_model.py:
```diff
@@ -1,11 +1,1323 @@
+        self.input_bounds = None  # bounds for the test input values
+
+    @property
+    def has_gradient(self):
```
Could you add a bit comment on what this function is actually checking?

amal-ghamdi replied:
done

chaozg reviewed (approved):
LGTM. Thanks @amal-ghamdi .
